### ISSUE TITLE
Editing a page with a form containing a query clobbers properties

### ISSUE THREAD

**Issue by cicalese** [SemanticMediaWiki/SemanticMediaWiki#1131]
I’ve encountered an issue that I think is a bug (or a failure to consider context) in Semantic MediaWiki that manifests itself with Semantic Forms. If you put an #ask or #show query in a form definition, editing a page with the form will wipe out all non-special properties for the page. This happens as soon as you visit the page with action=formedit in the process of rendering the form. If you save the page, the properties are restored, but if instead you cancel or navigate away from the page while editing it with the form, the properties are gone and a Has query property associated with the query on the form appears. Special properties (e.g. Page creator, Page author, Modification date, Creation date) are unaffected. Purging the page restores the property values.

My environment:
- MediaWiki 1.25.1
- PHP 5.5.28
- MySQL 5.5.45
- Semantic MediaWiki 2.3 alpha (9758ed9)
- Semantic Forms 3.3.2
- Semantic Extra Special Properties 1.3.0
- quite a few other extensions I can list if they become relevant

I was able to reproduce this behavior on two different wikis on two different servers using the same software versions but different wiki content. The only observed difference was that on the second wiki, the properties “healed themselves” after a few seconds, presumably because of a background job, but they did not ever get corrected without a purge or page save on the first wiki.

My assumption is that the code that adds the Has query properties is somehow stomping on the other properties.

**Comment by mwjames** [comment]
> I’ve encountered an issue that I think is a bug (or a failure to consider context) in Semantic MediaWiki that manifests itself with Semantic Forms. 

If it only can be observed in connection with SF then this is not a SMW issue, on the other-hand if it can be reproduced without SF then this can be categorised as a SMW bug (PS: I don't use nor have I installed SF and I haven't seen the issue).

**Comment by mwjames** [comment]
> My assumption is that the code that adds the Has query properties is somehow stomping on the other properties.

`Has query` [0] is being processed within the same transaction that follows the request for the query result and is being attached to the `ParserOutput` that is used during the update in [1].

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/includes/parserhooks/AskParserFunction.php#L200-L216

[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/LinksUpdateConstructed.php#L46-L64

**Comment by mwjames** [comment]
@kghbln Can this be observed without SF?

**Comment by cicalese** [comment]
I would counter that SMW should not write properties when action=formedit or most other actions. This problem likely is not observed with other actions because other actions do not execute queries.

**Comment by mwjames** [comment]
> I would counter that SMW should not write properties when action=formedit or most other actions.

SMW doesn't know anything about `action=formedit` (nor should it be), SMW only initiates an `ParserData::updateStore` (this when the `SemanticData` starts the processing by `SQLStore` / `SPARQLStore`) during:
- `SMW\UpdateJob`
- on the `LinksUpdateConstructed` hook event
- on the `FileUpload` hook event
- on the `ArticlePurge` hook event (executed in the `ParserAfterTidy` hook)

All data are expected to be added to the `ParserOutput` object with `ParserData` being the handler class to manage the in-memory `SemanticData` while in transaction.

Tests (existing unit and integration [0, 1]) have not shown any divergent behaviour (we test without any SF) for 1.25 / SMW 2.3 in regards to property annotations.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/tree/master/tests/phpunit/Integration/Parser

[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/tree/master/tests/phpunit/Integration/Query/Fixtures

**Comment by cicalese** [comment]
I understand that SMW is not aware of the current action. I did a bit more digging to figure out what is going on. SemanticForms is calling Parser::parse() on the content of the form (SF_FormUtils.php line 439). In the parser, the ParserAfterTidy hook is calling Semantic MediaWIkis's ParserData::updateStore(). The parser is associated with the page being editing, so updateStore is wiping out the properties on that page, but it should not. If I instantiate a new Parser object pointing to a dummy page Title, the properties on the page being editing do not get wiped out. There are two problems with that approach: 1) I'm not sure of the correct way to create a dummy page Title, so this could clobber a random page and 2) since the parser thinks it is parsing a different page, self-links on the page are not handled properly. Is there another way to invoke the parser so that it will be in the context of the correct page but will not clobber the properties?

**Comment by mwjames** [comment]
> In the parser, the ParserAfterTidy hook is calling Semantic MediaWIkis's ParserData::updateStore(). The parser is associated with the page being editing, so updateStore is wiping out the properties on that page,

`ParserData::updateStore` is only called for [0] when the ArticlePurge hook was called with a key to match `$cache->contains( $key )` (we have to use the workaround because `ArticlePurge` runs in a different transaction) hence a normal `Parser::parse` should not suppose to enter this tree.

If you want know whether the tree is executed or not, add something like `wfDebug( __METHOD__ . wfGetCaller( 3 ) )` before the  `ParserData::updateStore` call.

PS: You have to adjust the level of `wfGetCaller` to see the origin caller.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/ParserAfterTidy.php#L135-L138

**Comment by mwjames** [comment]
@cicalese Here is a quick schematic that describes the relationship between Parse, ParserOutput, ParserData, and the SemanticData object.

![image](https://cloud.githubusercontent.com/assets/1245473/9608489/e38026ec-50ce-11e5-9239-46fa42a8b413.png)

**Comment by cicalese** [comment]
Calling wfGetAllCallers() in ParserData::updateStore(), here is the tail of the call stack from the function calling Parser::parse() that I mentioned above:

SFFormUtils::getFormDefinition
Parser::parse
Parser::internalParseHalfParsed
Hooks::run/call_user_func_array
SMW\MediaWiki\Hooks\{closure}
SMW\MediaWiki\Hooks\ParserAfterTidy::process
SMW\MediaWiki\Hooks\ParserAfterTidy::performUpdate
SMW\MediaWiki\Hooks\ParserAfterTidy::checkForRequestedUpdateByPagePurge
SMW\ParserData::updateStore

**Comment by mwjames** [comment]
What is the `$this->parser->getTitle()->getArticleID()` when `ParserData::updateStore` is executed?

**Comment by cicalese** [comment]
$this->parser is undefined in ParserData::updateStore().

In SFFormUtils::getFormDefinition before the call to $parser->parse(), $parser->getTitle()->getArticleID() = 580

In SMW\Parserdata::updateStore, $this->title->getArticleID() = 580

**Comment by cicalese** [comment]
And, checking the page table in the database for page_id=580, this refers to the page that is being edited.

**Comment by mwjames** [comment]
OK, the only way for 580 to match the `if( $cache->contains( $key ) )` condition is when [0] was somehow triggered before `PaserAfterTidy` for the 580 ArticleID.

If it was matched it is gets deleted as soon as `ParserData::updateStore` is triggered.

What happens when you add `if( $cache->contains( $key ) && $cache->fetch( $key ) ) {`

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/ArticlePurge.php#L39-L44

**Comment by mwjames** [comment]
> $this->parser is undefined in ParserData::updateStore()

Just a second, $this->parser is undefined?

What is the result of `$this->parser->getTitle()->getArticleID()`? It cannot be undefined as it is required to create a $key, otherwise `$this->parser->getTitle()` would return with an error of `access on object null`.

```
		$key = $this->applicationFactory->newCacheFactory()->getPurgeCacheKey(
			$this->parser->getTitle()->getArticleID()
		);
```

**Comment by cicalese** [comment]
I do not believe it is ever getting into ArticlePurge::process(). I added the if statement you requested around the $pageId > 0 if statement in that function, and it made no difference. I added a die() before the if statement in a brute force attempt to see if it was getting there, and it did not die.

**Comment by cicalese** [comment]
You're talking about ParserData.php in the includes directory, right? There is no $this->parser in that file. I cannot find the lines of code that you are referring to.

**Comment by cicalese** [comment]
OK, your previous reference was to ArticlePurge. I will try that.

**Comment by mwjames** [comment]
> You're talking about ParserData.php in the includes directory, right? There is no $this->parser in that file. I cannot find the lines of code that you are referring to.

No, `ParserAfterTidy`.

`ParserData` is just a container but I'd like to know why [0] in `ParserAfterTidy` is entered as is not suppose to when not triggered by `ArticlePurge`.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/ParserAfterTidy.php#L135

**Comment by cicalese** [comment]
OK, changing the if statement in ParserAfterTidy fixed the problem!

**Comment by cicalese** [comment]
So, is that the fix, or is it a symptom of something else going on?

**Comment by mwjames** [comment]
> I added a die() before the if statement in a brute force attempt to see if it was getting there, and it did not die

I couldn't find any other explanation and as precautionary measure we check for TRUE explicitly as it is what we demand from `ArticlePurge` to set as value.

I also added an additional unit test case to check against FALSE [0].

> So, is that the fix, or is it a symptom of something else going on?

Just merged #1132.

It hasn't emerged in any of our core tests therefore I'd say it should be fixed for when `Parse:parse` is arbitrarily called.

Happy testing.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/1132/files#diff-cb2c3e06ec4bbdead86d69a7aff8381bR285

**Comment by cicalese** [comment]
Thank you!
